src/core/ext/filters/client_channel/lb_policy/grpclb/client_load_reporting_filter.cc

```diff
@@ -33,118 +33,49 @@
-    grpc_call_element* elem, const grpc_call_element_args* args) {
-  GPR_ASSERT(args->context != nullptr);
-  new (elem->call_data) call_data();
-  return absl::OkStatus();
-}
-
-static void clr_destroy_call_elem(grpc_call_element* elem,
-                                  const grpc_call_final_info* /*final_info*/,
-                                  grpc_closure* /*ignored*/) {
-  call_data* calld = static_cast<call_data*>(elem->call_data);
-  if (calld->client_stats != nullptr) {
-    // Record call finished, optionally setting client_failed_to_send and
-    // received.
-    calld->client_stats->AddCallFinished(
-        !calld->send_initial_metadata_succeeded /* client_failed_to_send */,
-        calld->recv_initial_metadata_succeeded /* known_received */);
+ArenaPromise<ServerMetadataHandle> ClientLoadReportingFilter::MakeCallPromise(
+    CallArgs call_args, NextPromiseFactory next_promise_factory) {
+  // Stats object to update.
+  RefCountedPtr<GrpcLbClientStats> client_stats;
+
+  // Handle client initial metadata.
+  // Grab client stats object from metadata.
+  auto client_stats_md =
+      call_args.client_initial_metadata->Take(GrpcLbClientStatsMetadata());
+  if (client_stats_md.has_value()) {
+    client_stats.reset(*client_stats_md);
   }
-  calld->~call_data();
-}
 
-static void clr_start_transport_stream_op_batch(
-    grpc_call_element* elem, grpc_transport_stream_op_batch* batch) {
-  call_data* calld = static_cast<call_data*>(elem->call_data);
-  // Handle send_initial_metadata.
-  if (batch->send_initial_metadata) {
-    // Grab client stats object from metadata.
-    auto client_stats_md =
-        batch->payload->send_initial_metadata.send_initial_metadata->Take(
-            grpc_core::GrpcLbClientStatsMetadata());
-    if (client_stats_md.has_value()) {
-      grpc_core::GrpcLbClientStats* client_stats = *client_stats_md;
-      if (client_stats != nullptr) {
-        calld->client_stats.reset(client_stats);
-        // Intercept completion.
-        calld->original_on_complete_for_send = batch->on_complete;
-        GRPC_CLOSURE_INIT(&calld->on_complete_for_send, on_complete_for_send,
-                          calld, grpc_schedule_on_exec_ctx);
-        batch->on_complete = &calld->on_complete_for_send;
-      }
-    }
-  }
-  // Intercept completion of recv_initial_metadata.
-  if (batch->recv_initial_metadata) {
-    calld->original_recv_initial_metadata_ready =
-        batch->payload->recv_initial_metadata.recv_initial_metadata_ready;
-    GRPC_CLOSURE_INIT(&calld->recv_initial_metadata_ready,
-                      recv_initial_metadata_ready, calld,
-                      grpc_schedule_on_exec_ctx);
-    batch->payload->recv_initial_metadata.recv_initial_metadata_ready =
-        &calld->recv_initial_metadata_ready;
-  }
-  // Chain to next filter.
-  grpc_call_next_op(elem, batch);
+  auto* server_initial_metadata = call_args.server_initial_metadata;
+
+  return Seq(next_promise_factory(std::move(call_args)),
```

Comment:
Wondering why `next_promise_factory` is put before everything while `grpc_call_next_op()` is put at the end?

Reply:
They're a little different:
- `start_transport_stream_op_batch` is called once per batch of operations, so we do our work on that batch and then we pass control to the next one with `grpc_call_next_op` -- in some other filters we might also delay that call for some batches until some async work completes.
- `MakeCallPromise` is called once per call and constructs a promise that will do the rest of the work. Since there's no asynchronous work to do to complete initial metadata for this filter we can immediately call the constructor for the next filter's promise as part of this one, and combine it into the total structure we return.